api/pkg/transformer/types/table/table_converter.go

```diff
@@ -24,8 +31,182 @@ func TableToJson(tbl *Table, format spec.FromTable_JsonFormat) (interface{}, err
+	}
+	return rowIDValues
+}
+
+func convertToUPIColumns(cols []*series.Series) ([]*upiv1.Column, error) {
+	upiCols := make([]*upiv1.Column, len(cols))
+	for idx, col := range cols {
+		colName := col.Series().Name
+		upiColType := upiv1.Type_TYPE_STRING
+		switch col.Type() {
+		case series.Int:
+			upiColType = upiv1.Type_TYPE_INTEGER
+		case series.Float:
+			upiColType = upiv1.Type_TYPE_DOUBLE
+		case series.String:
+			upiColType = upiv1.Type_TYPE_STRING
+		default:
+			return nil, fmt.Errorf("type %v is not supported in UPI", col.Type())
+		}
+
+		upiCols[idx] = &upiv1.Column{
+			Name: colName,
+			Type: upiColType,
+		}
+	}
+	return upiCols, nil
+}
+
+// ToUPITable converts table into upi table
+// will exclude row_id columns from upi columns
+func ToUPITable(tbl *Table, name string) (*upiv1.Table, error) {
+	cols := tbl.ColumnsWithExcluding([]string{RowIDColumn})
+	upiCols, err := convertToUPIColumns(cols)
+	if err != nil {
+		return nil, err
+	}
+
+	rowIDValues := getRowIDValues(tbl)
+	colValues, err := getTableRecordsForColumns(cols)
+	if err != nil {
+		return nil, err
+	}
+
+	upiRows := make([]*upiv1.Row, tbl.NRow())
+	rowIDColExist := tbl.NRow() == len(rowIDValues)
+	for rowIdx := 0; rowIdx < tbl.NRow(); rowIdx++ {
+		vals := make([]*upiv1.Value, len(cols))
+		for colIdx, col := range upiCols {
+			colValues := colValues[col.Name]
+			colVal := colValues[rowIdx]
+			val := &upiv1.Value{}
+			if colVal == nil {
+				val.IsNull = true
+				vals[colIdx] = val
+				continue
+			}
+
+			switch col.Type {
+			case upiv1.Type_TYPE_INTEGER:
+				intVal, err := converter.ToInt64(colVal)
+				if err != nil {
+					return nil, err
+				}
+				val.IntegerValue = intVal
+			case upiv1.Type_TYPE_DOUBLE:
+				floatVal, err := converter.ToFloat64(colVal)
+				if err != nil {
+					return nil, err
+				}
+				val.DoubleValue = floatVal
+			case upiv1.Type_TYPE_STRING:
+				strVal, err := converter.ToString(colVal)
+				if err != nil {
+					return nil, err
+				}
+				val.StringValue = strVal
+			default:
+				return nil, fmt.Errorf("not supported type")
+			}
+			vals[colIdx] = val
+		}
+		var rowID string
+		if rowIDColExist {
+			rowID = rowIDValues[rowIdx]
+		}
+		rowVal := &upiv1.Row{
+			RowId:  rowID,
+			Values: vals,
+		}
+		upiRows[rowIdx] = rowVal
+	}
+
+	upiTbl := &upiv1.Table{
+		Name:    name,
+		Columns: upiCols,
+		Rows:    upiRows,
+	}
+	return upiTbl, nil
+}
+
+// FromUPITable convert UPI table into standard transformer table
+func FromUPITable(tbl *upiv1.Table) (*Table, error) {
+	cols := tbl.Columns
+	lastColsIdx := len(cols)
+	colTypeLookup := make(map[int]upiv1.Type)
+	columnNames := make([]string, lastColsIdx+1)
+	for i, col := range cols {
+		colTypeLookup[i] = col.Type
+		columnNames[i] = col.Name
+	}
+	// adding row_id column
+	colTypeLookup[lastColsIdx] = upiv1.Type_TYPE_STRING
+	columnNames[lastColsIdx] = RowIDColumn
+
+	getValFn := func(value *upiv1.Value, cType upiv1.Type) (any, error) {
```

Comment:
Can it be extracted outside this function? It makes the function long and harder to understand.